File: src/assistir.js
const { chromium } = require('playwright');

const EMAIL = process.env.EMAIL;
const PASSWORD = process.env.PASSWORD;

if (!EMAIL || !PASSWORD) {
  console.error('::error title=Credenciais ausentes::Defina EMAIL e PASSWORD como secrets/variáveis.');
  process.exit(1);
}

function groupStart(title) { console.log(`::group::${title}`); }
function groupEnd() { console.log('::endgroup::'); }
function notice(msg) { console.log(`::notice title=Info::${msg}`); }
function warn(msg) { console.log(`::warning title=Atenção::${msg}`); }
function fail(msg) { console.log(`::error title=Erro::${msg}`); }

async function waitAndClick(page, selector, opts = {}) {
  const timeout = opts.timeout ?? 30000;
  await page.waitForSelector(selector, { timeout });
  await page.click(selector);
}

async function tryClickAvancar(page) {
  const byRole = page.getByRole('button', { name: /Avan\w*çar/i });
  if (await byRole.count()) { await byRole.first().click(); return true; }
  const bySpan = page.locator('button:has(span.MuiButton-label:has-text("Avançar"))');
  if (await bySpan.count()) { await bySpan.first().click(); return true; }
  const byText = page.locator('text=Avançar');
  if (await byText.count()) { await byText.first().click(); return true; }
  return false;
}

async function ensureLoggedIn(page) {
  groupStart('Login');
  notice('Abrindo página de login do Campus Digital');
  await page.goto('https://campusdigital.pucrs.br/login', { waitUntil: 'domcontentloaded' });
  notice('Clicando em "Entrar"');
  await waitAndClick(page, 'button:has-text("Entrar")');
  notice('Aguardando campo de e-mail');
  await page.waitForSelector('input[type="email"]', { timeout: 45000 });
  await page.fill('input[type="email"]', EMAIL);
  await page.click('input[type="submit"]');
  notice('Aguardando campo de senha');
  await page.waitForSelector('input[name="passwd"]', { timeout: 45000 });
  await page.fill('input[name="passwd"]', PASSWORD);
  await page.click('input[type="submit"]');
  notice('Tratando "Manter conectado?" se aparecer');
  try {
    await page.waitForSelector('#KmsiCheckboxField', { timeout: 15000 });
    await page.click('input[type="submit"]');
  } catch { warn('Tela "Manter conectado?" não apareceu (ok).'); }
  notice('Aguardando home do Campus Digital');
  await page.waitForURL('**/home', { timeout: 60000 });
  groupEnd();
}

async function openDisciplines(page) {
  groupStart('Navegar para disciplinas');
  notice('Clicando em "Ver Disciplinas"');
  await waitAndClick(page, 'button[data-cy="buttonSeeDisciplines"]');
  notice('Aguardando lista de disciplinas aparecer');
  await page.waitForSelector(
    'a.MuiTypography-root.MuiLink-root.MuiLink-underlineHover.MuiTypography-colorPrimary',
    { timeout: 60000 }
  );
  notice('Coletando os dois primeiros links de disciplinas');
  const links = await page.$$eval(
    'a.MuiTypography-root.MuiLink-root.MuiLink-underlineHover.MuiTypography-colorPrimary',
    els => els.slice(0, 2).map(e => e.href)
  );
  if (!links.length) throw new Error('Nenhuma disciplina encontrada.');
  console.log('Links encontrados:', links);
  groupEnd();
  return links.slice(0, 2);
}

function parseTimeToMs(timeStr) {
  const s = (timeStr || '').trim().toLowerCase();
  if (/^\d{1,2}:\d{2}(:\d{2})?$/.test(s)) {
    const parts = s.split(':').map(n => parseInt(n, 10));
    let seconds = 0;
    if (parts.length === 2) seconds = parts[0] * 60 + parts[1];
    else seconds = parts[0] * 3600 + parts[1] * 60 + parts[2];
    return (seconds + 60) * 1000;
  }
  let hours = 0, minutes = 0, seconds = 0;
  const hMatch = s.match(/(\d+)\s*h/);
  const mMatch = s.match(/(\d+)\s*m/);
  const secMatch = s.match(/(\d+)\s*s/);
  if (hMatch) hours = parseInt(hMatch[1], 10);
  if (mMatch) minutes = parseInt(mMatch[1], 10);
  if (secMatch) seconds = parseInt(secMatch[1], 10);
  if (hMatch || mMatch || secMatch) {
    const total = hours * 3600 + minutes * 60 + seconds;
    return (total + 60) * 1000;
  }
  return 5 * 60 * 1000;
}

async function logLessonContext(page) {
  try {
    const tempo = await page.locator('p[data-cy="timePartLesson"]').first().innerText({ timeout: 4000 });
    notice(`Tempo informado na página: ${tempo}`);
  } catch {}
  try {
    const info = await page.locator('div.infoCard').first().innerText({ timeout: 2000 });
    if (info && info.trim()) notice(`infoCard: ${info.trim()}`);
  } catch {}
  try {
    const part = await page.locator('div[data-cy="partLesson"]').first().innerText({ timeout: 2000 });
    if (part && part.trim()) notice(`partLesson: ${part.trim()}`);
  } catch {}
}

async function clickPlay(page) {
  const overlayPlay = page.locator('button[data-play-button="true"]').first();
  if (await overlayPlay.count()) {
    notice('Play overlay encontrado, clicando...');
    try {
      await overlayPlay.click({ timeout: 5000 });
      return true;
    } catch {}
  }
  const vimeoFrame = page.locator('iframe[src*="player.vimeo.com"]').first();
  if (await vimeoFrame.count()) {
    notice('Iframe do Vimeo encontrado; clicando no centro para iniciar.');
    try {
      await vimeoFrame.waitFor({ state: 'visible', timeout: 10000 });
      await vimeoFrame.scrollIntoViewIfNeeded();
      await vimeoFrame.click({ position: { x: 50, y: 50 } });
      await page.waitForTimeout(500);
      await vimeoFrame.click({ position: { x: 50, y: 50 } });
      return true;
    } catch {}
    try {
      notice('Tentando iniciar com tecla Espaço focando o iframe.');
      const frameHandle = await vimeoFrame.elementHandle();
      if (frameHandle) {
        await frameHandle.focus();
        await page.keyboard.press('Space');
        await page.waitForTimeout(500);
        return true;
      }
    } catch {}
  }
  warn('Não foi possível acionar o play.');
  return false;
}

async function playAndWaitForVideo(page) {
  await logLessonContext(page);
  let tempoMs;
  try {
    const tempoTexto = await page.locator('p[data-cy="timePartLesson"]').first().innerText({ timeout: 6000 });
    tempoMs = parseTimeToMs(tempoTexto.trim());
    notice(`Tempo da aula: ${tempoTexto} (+1 min) => aguardar ~${Math.round(tempoMs / 1000)}s`);
  } catch {
    warn('Não foi possível ler o tempo da aula. Usando 5min como fallback.');
    tempoMs = 5 * 60 * 1000;
  }
  const started = await clickPlay(page);
  if (!started) warn('Prosseguindo mesmo sem confirmação do play.');
  await page.waitForTimeout(1500);
  await page.waitForTimeout(tempoMs);
  return true;
}

async function processDisciplina(page, link, idx) {
  groupStart(`Disciplina ${idx + 1}`);
  notice(`Abrindo disciplina: ${link}`);
  await page.goto(link, { waitUntil: 'domcontentloaded' });
  let passos = 0;
  while (true) {
    passos++;
    groupStart(`Aula/Página ${passos}`);
    await playAndWait
        await playAndWaitForVideo(page);

    const avancou = await tryClickAvancar(page);
    if (!avancou) {
      notice('Botão "Avançar" não existe mais. Fim da disciplina.');
      groupEnd();
      break;
    }
    notice('Avançando para a próxima página/aula.');
    await page.waitForLoadState('domcontentloaded', { timeout: 60000 });

    groupEnd();
  }
  groupEnd();
}

(async () => {
  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext({
    viewport: { width: 1366, height: 768 },
    recordVideo: { dir: 'session-videos', size: { width: 1280, height: 720 } }
  });
  const page = await context.newPage();
  await context.tracing.start({ screenshots: true, snapshots: true, sources: true });

  try {
    await ensureLoggedIn(page);
    const links = await openDisciplines(page);

    // Primeira disciplina
    await processDisciplina(page, links[0], 0);

    // Voltar para lista e abrir segunda disciplina
    groupStart('Retornando à lista de disciplinas');
    await page.goto('https://campusdigital.pucrs.br/home', { waitUntil: 'domcontentloaded' });
    await waitAndClick(page, 'button[data-cy="buttonSeeDisciplines"]');
    await page.waitForSelector(
      'a.MuiTypography-root.MuiLink-root.MuiLink-underlineHover.MuiTypography-colorPrimary',
      { timeout: 60000 }
    );
    groupEnd();

    const linksNovos = await page.$$eval(
      'a.MuiTypography-root.MuiLink-root.MuiLink-underlineHover.MuiTypography-colorPrimary',
      els => els.slice(0, 2).map(e => e.href)
    );

    await processDisciplina(page, linksNovos[1], 1);

    notice('Todas as aulas das duas disciplinas foram processadas com sucesso.');
  } catch (err) {
    fail(`Falha na automação: ${err.message}`);
    process.exitCode = 1;
  } finally {
    await context.tracing.stop({ path: 'playwright-trace.zip' });
    await context.close();
    await browser.close();
  }
})();

File: assistir.py
from playwright.sync_api import sync_playwright
import os
import re
import time

# Funções de log no padrão GitHub Actions
def group_start(title):
    print(f"::group::{title}")

def group_end():
    print("::endgroup::")

def notice(msg):
    print(f"::notice title=Info::{msg}")

def warn(msg):
    print(f"::warning title=Atenção::{msg}")

def fail(msg):
    print(f"::error title=Erro::{msg}")

EMAIL = os.getenv("EMAIL")
PASSWORD = os.getenv("PASSWORD")

if not EMAIL or not PASSWORD:
    fail("Defina EMAIL e PASSWORD como variáveis de ambiente.")
    exit(1)

def parse_time_to_seconds(time_str):
    s = time_str.strip().lower()
    if re.match(r"^\d{1,2}:\d{2}(:\d{2})?$", s):
        parts = list(map(int, s.split(":")))
        if len(parts) == 2:
            seconds = parts[0] * 60 + parts[1]
        else:
            seconds = parts[0] * 3600 + parts[1] * 60 + parts[2]
        return seconds + 60
    h = re.search(r"(\d+)\s*h", s)
    m = re.search(r"(\d+)\s*m", s)
    sec = re.search(r"(\d+)\s*s", s)
    hours = int(h.group(1)) if h else 0
    minutes = int(m.group(1)) if m else 0
    seconds = int(sec.group(1)) if sec else 0
    if h or m or sec:
        return hours * 3600 + minutes * 60 + seconds + 60
    return 5 * 60

def click_play(page):
    overlay = page.locator('button[data-play-button="true"]').first
    if overlay.count():
        try:
            notice("Play overlay encontrado, clicando...")
            overlay.click(timeout=3000)
            return True
        except:
            pass
    vimeo = page.locator('iframe[src*="player.vimeo.com"]').first
    if vimeo.count():
        notice("Iframe do Vimeo encontrado; clicando no centro para iniciar.")
        vimeo.wait_for(state="visible", timeout=10000)
        box = vimeo.bounding_box()
        if box:
            page.mouse.click(box["x"] + box["width"]/2, box["y"] + box["height"]/2)
            time.sleep(0.5)
            page.mouse.click(box["x"] + box["width"]/2, box["y"] + box["height"]/2)
            return True
    warn("Não foi possível acionar o play.")
    return False

def assistir_disciplina(page, link):
    group_start(f"Disciplina: {link}")
    page.goto(link)
    passo = 0
    while True:
        passo += 1
        group_start(f"Aula/Página {passo}")
        try:
            tempo_txt = page.locator('p[data-cy="timePartLesson"]').first.inner_text(timeout=5000)
            tempo_seg = parse_time_to_seconds(tempo_txt)
            notice(f"Tempo da aula: {tempo_txt} (+1 min) => aguardando {tempo_seg}s")
        except:
            tempo_seg = 5 * 60
            warn("Tempo não encontrado, usando 5min como fallback.")

        started = click_play(page)
        if not started:
            warn("Prosseguindo mesmo sem confirmação do play.")

        time.sleep(tempo_seg)

        avancar = page.get_by_role("button", name=re.compile("Avançar", re.I))
        if avancar.count():
            notice("Avançando para a próxima página/aula.")
            avancar.first.click()
            page.wait_for_load_state("domcontentloaded")
        else:
            notice("Fim da disciplina.")
            group_end()
            break
        group_end()
    group_end()

with sync_playwright() as p:
    group_start("Iniciando script")
    browser = p.chromium.launch(headless=True)
    context = browser.new_context(viewport={"width": 1366, "height": 768})
    page = context.new_page()

    group_start("Login")
    notice("Abrindo página de login do Campus Digital")
    page.goto("https://campusdigital.pucrs.br/login")
    notice('Clicando em "Entrar"')
    page.click('button:has-text("Entrar")')
    notice("Aguardando campo de e-mail")
    page.fill('input[type="email"]', EMAIL)
    page.click('input[type="submit"]')
    notice("Aguardando campo de senha")
    page.fill('input[name="passwd"]', PASSWORD)
    page.click('input[type="submit"]')
    notice('Tratando "Manter conectado?" se aparecer')
    try:
        page.wait_for_selector('#KmsiCheckboxField', timeout=5000)
        page.click('input[type="submit"]')
    except:
        warn('Tela "Manter conectado?" não apareceu (ok).')
    page.wait_for_url("**/home")
    group_end()

    group_start("Navegar para disciplinas")
    notice('Clicando em "Ver Disciplinas"')
    page.click('button[data-cy="buttonSeeDisciplines"]')
    notice("Aguardando lista de disciplinas aparecer")
    page.wait_for_selector('a.MuiTypography-root.MuiLink-root')
    links = page.eval_on_selector_all(
        'a.MuiTypography-root.MuiLink-root',
        'els => els.slice(0, 2).map(e => e.href)'
    )
    notice(f"Links encontrados: {links}")
    group_end()

    assistir_disciplina(page, links[0])

    group_start("Retornando à lista de disciplinas")
    page.goto("https://campusdigital.pucrs.br/home")
    page.click('button[data-cy="buttonSeeDisciplines"]')
    page.wait_for_selector('a.MuiTypography-root.MuiLink-root')
    links2 = page.eval_on_selector_all(
        'a.MuiTypography-root.MuiLink-root',
        'els => els.slice(0, 2).map(e => e.href)'
    )
    group_end()

    assistir_disciplina(page, links2[1])

    browser.close()
    group_end()
